core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -167,13 +175,96 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
+  def createSnapshotsByWorkspaceIdV3(workspaceId: String, snapshotIds: Set[UUID]): Future[Unit] =
+    getV2WorkspaceContextAndPermissionsById(workspaceId,
+                                            SamWorkspaceActions.write,
+                                            Some(WorkspaceAttributeSpecs(all = false))
+    ).flatMap { rawlsWorkspace =>
+      createSnapshots(rawlsWorkspace, snapshotIds)
+    }
+
+  // Find a workspace using the workspaceName then calls the createSnapshot method
+  def createSnapshotsByWorkspaceNameV3(workspaceName: WorkspaceName, snapshotIds: Set[UUID]): Future[Unit] =
+    getV2WorkspaceContextAndPermissions(workspaceName,
+                                        SamWorkspaceActions.write,
+                                        Some(WorkspaceAttributeSpecs(all = false))
+    ).flatMap(rawlsWorkspace => createSnapshots(rawlsWorkspace, snapshotIds))
+
+  // Link the snapshot pao to the workspace pao
+  private def createSnapshots(rawlsWorkspace: Workspace, snapshotIds: Set[UUID]): Future[Unit] =
+    for {
+      snapshotsFromDataRepo <- Future {
+        snapshotIds.map(getSnapshotFromDataRepoWithId)
+      }
+      workspacePaoOpt <- policyService.getPao(rawlsWorkspace.workspaceIdAsUUID, ctx)
+      workspacePao = workspacePaoOpt.getOrElse(
+        throw new RawlsExceptionWithErrorReport(
+          ErrorReport(StatusCodes.NotFound,
+                      s"Workspace PAO not found for workspace ${rawlsWorkspace.workspaceIdAsUUID}"
+          )
+        )
+      )
+
+      snapshotPaos <- Future.traverse(snapshotsFromDataRepo) { snapshot =>
+        policyService.getOrCreateSnapshotPao(snapshot.getId, ctx)
+      }
+
+      _ <- Future.traverse(snapshotsFromDataRepo) { snapshot =>
+        policyService.linkSnapshotPaoToWorkspacePao(snapshot.getId,
+                                                    rawlsWorkspace.workspaceIdAsUUID,
+                                                    dryRun = true,
+                                                    ctx
+        )
+      }
+
+      // if any snapshots contain protected data, the workspace must be protected
+      _ = if (
+        snapshotPaos.exists(PolicyUtilities.containsProtectedDataPolicy) && !PolicyUtilities
+          .containsProtectedDataPolicy(workspacePao)
+      ) {
+        throw new ProtectedDataException("Unable to add protected snapshot to unprotected workspace.")
+      }
+
+      // Region constraints can cause conflicts when combining PAOs. We check that no individual
+      // snapshot PAO will conflict with the workspace PAO above, but there's no easy way to check
+      // that all of the snapshot PAOs will combine together cleanly when they're all linked to
+      // the same workspace. Snapshots shouldn't have region constraint policies, but throw here
+      // just in case.
+      _ = if (snapshotPaos.exists(PolicyUtilities.containsRegionConstraintPolicy)) {
```

Comment:
I am also not convinced that this is the right approach. It's a cheap check, so it's not too much trouble, but we don't expect any snapshot or workspace PAOs to include a `region-constraint` policy.  Even if there were a `region-constraint` policy, it's possible that the linking would succeed and we'd be preventing that here unnecessarily. Feedback welcome!